nav2_behavior_tree/plugins/decorator/planner_selector_node.cpp

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) 2018 Intel Corporation
+// Copyright (c) 2020 Pablo Iñigo Blasco
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <string>
+#include <memory>
+
+#include "std_msgs/msg/string.hpp"
+#include "behaviortree_cpp_v3/decorator_node.h"
+
+#include "nav2_behavior_tree/plugins/decorator/planner_selector_node.hpp"
+
+#include "rclcpp/rclcpp.hpp"
+
+namespace nav2_behavior_tree
+{
+
+using std::placeholders::_1;
+
+PlannerSelector::PlannerSelector(
+  const std::string & name,
+  const BT::NodeConfiguration & conf)
+: BT::DecoratorNode(name, conf)
+{
+  auto node = config().blackboard->get<rclcpp::Node::SharedPtr>("node");
```

Comment:
Should you spin_some in the main processing function so the node can process callbacks?

Reply:
I do not know exactly the origin of node, I think we do not need to spin it in this class.

My impression was that was the main node of the bt_navigator_node process.
I think it is being "spinned" somewhere, but I can check.

Reply:
There's no spinning of this node anywhere outside of the behavior tree, its only spun for `spin_until_XYZ_complete` and `spin_some` for actions in the FollowPath / ComputePathToPose nodes. So without spinning the node a little here, you're relying on those 2 nodes to be in the BT and processing to use this feature. 

That might be OK? I don't think it breaks anything to not include it here, assuming you have those other 2 nodes processed regularly in your BT.

Reply:
Isnt that node spinned here?

https://github.com/ros-planning/navigation2/blob/main/nav2_behavior_tree/include/nav2_behavior_tree/bt_action_node.hpp#L157

Reply:
Then I think we should some call spin_some into the tick function before setting the selected_planner output port value.

I think would guarantee that the node callbacks are handled at least once each time the BT passes through the planner_selector and at least the last ros message is handled.

Reply:
Agreed

Reply:
https://github.com/ros-planning/navigation2/issues/2251

Reply:
Having callback_group_executor_ sounds good.

This issue is related with my original question about how the nav-bt were supposed to interact with the external ros network receiving custom information/commands. So it looks these are the first times this sort of issues are being faced. It is an interesting issue, I will keep an eye, but no doubt there must be an optimal solution (maybe that one you say).

In this case, do you think it would be a problem if one message from other btnode is handled?
For example? would be a problem if a new goal message callback is handled? I would say no.

I think we can have an initial implementation based on rclcpp::spin_some and later move to a more advanced approach. What do you think?

Reply:
>  For example, would be a problem if a new goal message callback is handled? I would say no.

No, I don't think there's likely to be any "real" problem with them all processed together lump sum. If a message from the goal updater or another subscriber of "reasonable" update rate is processed when `spin_some`, I don't think there would be any noticable difference in the system.

I think the separation using the different executors though is a way to verifiably guarantee there can't and won't be an issue, even in some odd edge case (like if someone implements a stupidly long service server in the BT node for some reason or decides to subscribe to a 10kHz topic that is washing out everything else).

Reply:
I agree, lets use the power of ROS2.

Reply:
@pabloinigoblasco lets not do it in this PR. It's come to my attention that the executors might be creating a new context remapping to a new DDS participant. This would not be good if we had each BT node creating a new participant to process separately. Some more research has to be done to figure out if this is workable before we jump into it.